### PR TITLE
Signal enum

### PR DESCRIPTION
**Pull request by fiveop**
This is work in progress. I post this pull request as a request for discussion. I mark the lines I have doubts or ideas about.

Ignore the first commit. It is the same as PR #361, which I wanted to base this change off.

### CODE REVIEW

File: src/sys/signal.rs [hunk 1]
```diff
@@ -6,44 +6,125 @@ use {Errno, Result};
...
-    SIGWINCH,
-    SIGUSR1,
-    SIGUSR2,
-};
-
-// This doesn't always exist, but when it does, it's 7
-pub const SIGEMT: libc::c_int = 7;
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Signal {
+    SIGHUP = libc::SIGHUP as isize,
```
Comment (fiveop) [inline src/sys/signal.rs]: I would have prefered `#[repr(libc::c_int)]` but that does not seem to be possible. Maybe that's worth an RFC for the language.

File: src/sys/signal.rs [hunk 2]
```diff
@@ -6,44 +6,125 @@ use {Errno, Result};
...
+    SIGIO = libc::SIGIO as isize,
+    #[cfg(not(target_os = "macos"))]
+    SIGPWR = libc::SIGPWR as isize,
+    SIGSYS = libc::SIGSYS as isize,
+    #[cfg(target_os = "macos")]
+    SIGEMT = libc::SIGEMT as isize,
+    #[cfg(target_os = "macos")]
+    SIGINFO = libc::SIGEMT as isize,
+}
+
```
Comment (fiveop) [inline src/sys/signal.rs]: The following type and function are only needed internally. They are used in the implementation of the iterator and in `wait.rs`. Because of the latter I hade to make them public (for now).

I propose putting the whole code into a private module (src/internal/sys/signal.rs) and then reexport what is supposed to be public here.

File: src/sys/signal.rs [hunk 3]
```diff
@@ -6,44 +6,125 @@ use {Errno, Result};
...
+pub const SIGUNUSED : Signal = SIGSYS;
 
 pub const NSIG: libc::c_int = 32;
 
+struct SignalIterator {
+    signum : libc::c_int,
+}
+
+impl Iterator for SignalIterator {
+    type Item = Signal;
```
Comment (fiveop) [inline src/sys/signal.rs]: I considered a nother method of iterating over enumerations. It uses an array of all elements and uses the slice module to produce and iterator. I think the following is a little clearer than an array which contains all the elements of the enumeration again. In particular, when the elements of the enumeration differ between platforms.

File: src/sys/signal.rs [hunk 4]
```diff
@@ -6,44 +6,125 @@ use {Errno, Result};
...
+            self.signum += 1;
+            Some(signal_from(self.signum))
+        } else {
+            None
+        }
+    }
+}
+
+impl Signal {
+    fn iterator() -> SignalIterator {
```
Comment (fiveop) [inline src/sys/signal.rs]: This should probably be public.

File: src/sys/signal.rs [hunk 5]
```diff
@@ -86,40 +166,33 @@ impl SigSet {
         SigSet { sigset: sigset }
     }
 
-    pub fn add(&mut self, signum: SigNum) -> Result<()> {
-        let res = unsafe { libc::sigaddset(&mut self.sigset as *mut libc::sigset_t, signum) };
-
-        Errno::result(res).map(drop)
+    pub fn add(&mut self, signal: Signal) {
```
Comment (utkarshkukreti) [inline src/sys/signal.rs]: I think we should rename this to `insert` to match the *Map and *Set APIs.

Comment (fiveop) [inline src/sys/signal.rs]: We could do that. Though, I would want to do that in a separate PR

**Comment by fiveop** [comment]
@kamalmarhubi I commented on yours. Now, it is your turn ;)

File: src/sys/signal.rs (hunk 1 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: I came across the same issue in #364 though I decided to go with `#[repr(i32)]` since the type of the constants in libc crate (`c_int` is `i32` unconditionally), and cast to `c_int` at use. Thinking about it a bit more, I'm not super happy about it, because if libc ever makes `c_int` conditional, we'll have trouble.

Comment (kamalmarhubi) [inline src/sys/signal.rs]: And yeah, I think it would be good if the language supported type aliases in `#[repr]` attributes.

File: src/sys/signal.rs [hunk 6]
```diff
@@ -6,44 +6,125 @@ use {Errno, Result};
...
+    SIGPWR = libc::SIGPWR as isize,
+    SIGSYS = libc::SIGSYS as isize,
+    #[cfg(target_os = "macos")]
+    SIGEMT = libc::SIGEMT as isize,
+    #[cfg(target_os = "macos")]
+    SIGINFO = libc::SIGEMT as isize,
+}
+
+// TODO: NOT PUBLIC :/
+pub type SigNum = libc::c_int;
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: I wonder if we could instead define `From<T> for Signal` for various integral types? Our code and user code could then write `Signal::from(9)` for example.

File: src/sys/signal.rs (hunk 2 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: For hiding, your proposal is the right thing to do. Though there's no need to put everything in that module I think: circular dependencies among modules in a crate are ok. You could have an internal module depend on this module, and define `signal_from`. Then this module and `wait` could both depend on that module.

File: src/sys/signal.rs (hunk 5 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: I'm ok with pushing that change to a later PR. This one can just be the type changing to an enum.

File: src/sys/signal.rs [hunk 7]
```diff
@@ -86,40 +166,33 @@ impl SigSet {
...
     }
 
-    pub fn contains(&self, signum: SigNum) -> Result<bool> {
-        let res = unsafe { libc::sigismember(&self.sigset as *const libc::sigset_t, signum) };
-
-        match try!(Errno::result(res)) {
-            1 => Ok(true),
-            0 => Ok(false),
-            _ => unreachable!("unexpected value from sigismember"),
+    pub fn extend(&mut self, other: &SigSet) {
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: later PR could then also make this take `FromIterator<Item=Signal>`, I like this :-)

File: src/sys/signal.rs (hunk 3 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: I think this approach is better.

File: src/sys/signal.rs (hunk 4 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: I wonder if it should be `all_signals` or something instead?

**Comment by kamalmarhubi** [comment]
Not for this PR: one thing from looking more closely at this module than I have before: do the thread functions really belong as `impl SigSet`? I think they would more belong on the thread object via some extensions trait.

File: src/sys/signal.rs (hunk 1 above)
Comment (fiveop) [inline src/sys/signal.rs]: I worte a draft for an RFC regarding this: https://github.com/fiveop/rfcs/blob/master/text/0000-type-aliases-in-enumeration-repr-attritubes.md

File: src/sys/signal.rs (hunk 2 above)
Comment (fiveop) [inline src/sys/signal.rs]: I'll implement that.

File: src/sys/signal.rs (hunk 3 above)
Comment (fiveop) [inline src/sys/signal.rs]: I have two problems with this: Its slower. It cannot be used with enumerations whose variant representation dont form a single interval in the natural numbers. Therefore, I actually prefer the other approach, even though it is more verbose and probably makes the executable slightly bigger.

File: src/sys/signal.rs (hunk 4 above)
Comment (fiveop) [inline src/sys/signal.rs]: I thought about the name as well, though I think iterator is pretty descriptive.

File: src/sys/signal.rs (hunk 6 above)
Comment (fiveop) [inline src/sys/signal.rs]: Signal::from(9) fits to the signal_from method above.In that case, I agree. However, I do not want to provide users access to it (as mentioned above).

File: src/sys/signal.rs (hunk 4 above)
Comment (utkarshkukreti) [inline src/sys/signal.rs]: `all_signals` would be a little redundant. How about `all`?

Comment (fiveop) [inline src/sys/signal.rs]: `all` (or `all_signals`) implies to me some sort of collection. Why not just use that as the name of the function. `iterator` returns an `Iterator`. `all` could also be an array or some other sort of collection.

File: src/sys/signal.rs (hunk 3 above)
Comment (fiveop) [inline src/sys/signal.rs]: After my last change it is no longer slower. However, we still cannot use its pattern for non consecutive enums.

**Comment by fiveop** [comment]
I worked on this PR again, after adapting code that uses the `sys::signal` module.

Since I have to write functions with C-ABI as signal handlers, which have to take `SigNum` instead of `Signal` because of type safety, I noticed that it does make sense to export the conversion function from `SigNum` to `Signal`, so the `From` trait fits the bill perfectly.

I am now pretty happy with the state of the PR. Any more remarks?

File: src/sys/signal.rs [hunk 8]
```diff
@@ -6,41 +6,92 @@ use {Errno, Result};
...
+    SIGPWR = libc::SIGPWR,
+    SIGSYS = libc::SIGSYS,
+    #[cfg(target_os = "macos")]
+    SIGEMT = libc::SIGEMT,
+    #[cfg(target_os = "macos")]
+    SIGINFO = libc::SIGEMT,
+}
+
+
+pub type SigNum = libc::c_int;
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: one thing I'm wondering is what to do we get out of an extra indirection here? posix specifies they be `int`:

> The <signal.h> header shall define the following macros that are used to refer to the signals that occur in the system. Signals defined here begin with the letters SIG followed by an uppercase letter. The macros shall expand to positive integer constant expressions with type int and distinct values. The value 0 is reserved for use as the null signal (see kill()). Additional implementation-defined signals may occur in the system.

—http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/signal.h.html

File: src/sys/signal.rs [hunk 9]
```diff
@@ -6,41 +6,92 @@ use {Errno, Result};
...
+    }
+}
+
+pub use self::Signal::*;
+
+pub const SIGIOT : Signal = SIGABRT;
+pub const SIGPOLL : Signal = SIGIO;
+pub const SIGUNUSED : Signal = SIGSYS;
 
 pub const NSIG: libc::c_int = 32;
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: is this correct on all platforms with the conditionally included variants?

File: src/sys/signal.rs (hunk 3 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: perhaps it's better to always have a static array of all the enum variants then?

File: src/sys/signal.rs (hunk 8 above)
Comment (fiveop) [inline src/sys/signal.rs]: It signals that we do not only want any int but one out of the range for signals.

File: src/sys/signal.rs (hunk 9 above)
Comment (fiveop) [inline src/sys/signal.rs]: Yes. That is what the other PR with the missing signals is for. I counted and hunted the missing ones down :)

File: src/sys/signal.rs (hunk 3 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: if the concern is returning a slice iterator, we could newtype it to make it opaque. that would simplify implementation and work for non-contiguous enums.

Comment (fiveop) [inline src/sys/signal.rs]: It probably is. I'll change it.

Comment (fiveop) [inline src/sys/signal.rs]: Why would returning a slice iterator be a problem?

File: src/sys/signal.rs (hunk 9 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: that seems annoying! I guess this is another place where a static array would help: NSIG would be `.len()` on it. There must be some generally accepted way of doing this!

Comment (fiveop) [inline src/sys/signal.rs]: `.len()` + 1, because there are actually 31 signals (+ the special 0 signal your quote above mentions :)).

File: src/sys/signal.rs (hunk 8 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: but a type alias doesn't do anything to help with that?

Comment (kamalmarhubi) [inline src/sys/signal.rs]: oh as documentation somehow, I see. hmmmm...

Comment (fiveop) [inline src/sys/signal.rs]: I admit, that one can argue, that it hides the fact that it's unsafe.

**Comment by fiveop** [comment]
I changed how the iterator is created.

File: src/sys/signal.rs (hunk 3 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: it's part of the public interface, so a change to the implementation would be breaking.

(my kingdom for an [`impl Trait`](https://github.com/rust-lang/rfcs/pull/1522)!)

Comment (fiveop) [inline src/sys/signal.rs]: So we should export our own iterator.

File: src/sys/signal.rs (hunk 8 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: I'm now wondering if we should be using `From` for this. From docs:

> **Note: this trait must not fail.** If the conversion can fail, use a dedicated method which return an `Option<T>` or a `Result<T, E>`.
―http://doc.rust-lang.org/std/convert/trait.From.html

Comment (utkarshkukreti) [inline src/sys/signal.rs]: The [TryFrom/TryInto RFC](https://github.com/rust-lang/rfcs/pull/1542) was accepted 2 days ago.

Comment (fiveop) [inline src/sys/signal.rs]: I will provide our own implementation and add a comment as well as issue regarding the new traits, once we merge this.

Comment (kamalmarhubi) [inline src/sys/signal.rs]: @utkarshkukreti amazing! I knew it was being worked on, hadn't realized it was already written and merged.

**Comment by fiveop** [comment]
I worked on this again. Two things changed:

* The `From` implementation was replaced by the `signal_from` function because `From` may not fail. A comment regarding the comming trait `TryFrom` was added.
* `Signal::iterator` now returns our own `Iterator` struct, so that changes in implementation won't break the interface.

**Comment by kamalmarhubi** [comment]
> * The `From` implementation was replaced by the `signal_from` function because `From` may not fail. A comment regarding the comming trait `TryFrom` was added.

Given this change, the motivation for having `SigNum` in our public API seems a bit weaker. Do you think we can remove it now?

**Comment by kamalmarhubi** [comment]
Also @fiveop thanks for being so patient through all the comments!

**Comment by fiveop** [comment]
I removed `SigNum`. With `signal_from` failing properly it makes sense.

Any more concerns?

File: src/lib.rs [hunk 10]
```diff
@@ -8,7 +8,6 @@
 // latest bitflags triggers a rustc bug with cross-crate macro expansions causing dead_code
 // warnings even though the macro expands into something with allow(dead_code)
 #![allow(dead_code)]
-#![deny(warnings)]
```
Comment (kamalmarhubi) [inline src/lib.rs]: is there a reason this was removed in this PR?

File: src/sys/signal.rs (hunk 9 above)
Comment (kamalmarhubi) [inline src/sys/signal.rs]: now that the iterator uses an internal array, should `NSIG` be public? should it ever have been public?

Comment (kamalmarhubi) [inline src/sys/signal.rs]: also if the `SIGNALS` array is moved up to module level, can this be `SIGNALS.len() + 1`, or is Rust's constexpr support not strong enough for that?

File: src/sys/signal.rs [hunk 11]
```diff
@@ -2,56 +2,162 @@
...
+    SIGEMT = libc::SIGEMT,
+    #[cfg(target_os = "macos")]
+    SIGINFO = libc::SIGEMT,
+}
+
+// We do not implement the From trait, because it is supposed to be infallible.
+// With Rust RFC 1542 comes the appropriate trait TryFrom. Once it is
+// implemented, we'll replace this function.
+#[inline]
+pub fn signal_from(signum: libc::c_int) -> Result<Signal> {
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: Could this perhaps be `Signal::from_c_int`?

**Comment by kamalmarhubi** [comment]
Ok I think I will promise not to make any more requests after these three I just made!

File: src/lib.rs (hunk 10 above)
Comment (fiveop) [inline src/lib.rs]: Probably, because it wouldn't compile otherwise on whatever was the current nightly then. I'll recheck it, when I look at it again.

File: src/sys/signal.rs (hunk 11 above)
Comment (fiveop) [inline src/sys/signal.rs]: Certainly

**Comment by fiveop** [comment]
I only see two messages? What's the third concern?

File: src/lib.rs (hunk 10 above)
Comment (kamalmarhubi) [inline src/lib.rs]: No worries, was just curious.

File: src/sys/signal.rs [hunk 12]
```diff
@@ -2,56 +2,162 @@
...
+    pub fn iterator() -> SignalIterator {
+        SignalIterator{next: 0}
+    }
+}
+
+pub use self::Signal::*;
+
+pub const SIGIOT : Signal = SIGABRT;
+pub const SIGPOLL : Signal = SIGIO;
+pub const SIGUNUSED : Signal = SIGSYS;
```
Comment (kamalmarhubi) [inline src/sys/signal.rs]: are these aliases cross-platform?

**Comment by kamalmarhubi** [comment]
@fiveop 
> I only see two messages? What's the third concern?

Honestly no idea! Maybe I can't count past one correctly. Anyway, feel free to r=me once `signal_from` changes. Thanks so much for doing this!

File: src/sys/signal.rs (hunk 9 above)
Comment (fiveop) [inline src/sys/signal.rs]: I still use NSIG in my code (in order to create an function array of NSIG size for handlers). I'll give the `len() + 1` definition a try.

Comment (fiveop) [inline src/sys/signal.rs]: And the answer is "No".

**Comment by fiveop** [comment]
finally done?

**Comment by kamalmarhubi** [comment]
@homu r+

:tada:

**Comment by kamalmarhubi** [comment]
@homu r-

**Comment by kamalmarhubi** [comment]
Some failures on Apple?

File: src/sys/signal.rs [hunk 13]
```diff
@@ -2,11 +2,60 @@
...
+    SIGPROF = libc::SIGPROF,
+    SIGWINCH = libc::SIGWINCH,
+    SIGIO = libc::SIGIO,
+    #[cfg(not(target_os = "macos"))]
+    SIGPWR = libc::SIGPWR,
+    SIGSYS = libc::SIGSYS,
+    #[cfg(target_os = "macos")]
+    SIGEMT = libc::SIGEMT,
+    #[cfg(target_os = "macos")]
+    SIGINFO = libc::SIGEMT,
```
Comment (utkarshkukreti) [inline src/sys/signal.rs]: @kamalmarhubi typo here?

Comment (utkarshkukreti) [inline src/sys/signal.rs]: cc @fiveop

**Comment by fiveop** [comment]
this should be good to merge now as well (finally! :))

**Comment by posborne** [comment]
Although @kamalmarhubi did most of the reviewing before, I looked things over again and it looks good to me.  @homu r+
